crates/storage/db/src/metrics.rs

```diff
@@ -1,10 +1,57 @@
+use crate::{Tables, NUM_TABLES};
+use dashmap::DashMap;
 use metrics::{Gauge, Histogram};
 use reth_libmdbx::CommitLatency;
 use reth_metrics::{metrics::Counter, Metrics};
-use std::time::{Duration, Instant};
+use rustc_hash::FxHasher;
+use std::{
+    hash::BuildHasherDefault,
+    str::FromStr,
+    time::{Duration, Instant},
+};
+use strum::EnumCount;
 
 const LARGE_VALUE_THRESHOLD_BYTES: usize = 4096;
 
+/// Caches metric handles for database environment to make sure handles are not re-created
+/// on every operation.
+#[derive(Debug)]
+pub struct DatabaseEnvMetrics {
+    /// Caches OperationMetrics handles for each table and operation tuple.
+    operations: DashMap<(Tables, Operation), OperationMetrics, BuildHasherDefault<FxHasher>>,
+}
+
+impl DatabaseEnvMetrics {
+    pub(crate) fn new() -> Self {
+        Self {
+            operations: DashMap::with_capacity_and_hasher(
+                NUM_TABLES * Operation::COUNT,
+                BuildHasherDefault::<FxHasher>::default(),
+            ),
+        }
+    }
+
+    /// Record a metric for database operation executed in `f`. Panics if the table name is unknown.
+    pub(crate) fn record_operation<T>(
+        &self,
+        table: &'static str,
```

Comment:
nit unrelated to this PR: maybe even no need for a separate `table` argument and we can use `T::NAME`?

Reply:
disregard, `T` is not a table here 🤦